### PR TITLE
Validate configuration ethereum.nodeURL

### PR DESCRIPTION
**Pull request by arberiii**
Hello 👋 , this is my first PR in go-centrifuge project. I picked [issue #1204](https://github.com/centrifuge/go-centrifuge/issues/1204) which is labeled as the good first issue and tried to work it out.

As described in the issue, etherNodeURL needs a prefix and by default should be set to https. The validateURL adds the prefix if necessary and checks if the schema is either http or https.

### CODE REVIEW

File: config/configuration.go [hunk 1]
```diff
@@ -651,7 +660,7 @@ func CreateConfigFile(args map[string]interface{}) (*viper.Viper, error) {
 	if p2pConnectTimeout != "" {
 		v.Set("p2p.connectTimeout", p2pConnectTimeout)
 	}
-	v.Set("ethereum.nodeURL", ethNodeURL)
+	v.Set("ethereum.nodeURL", validateURL(ethNodeURL))
 	v.Set("ethereum.accounts.main.key", "")
 	v.Set("ethereum.accounts.main.password", "")
 	v.Set("centChain.nodeURL", centChainURL)
```
Comment (vedhavyas) [inline config/configuration.go:666]: We would need to do the same check for the centrifugeChain URL as well

File: config/configuration.go [hunk 2]
```diff
@@ -683,3 +692,20 @@ func CreateConfigFile(args map[string]interface{}) (*viper.Viper, error) {
 func (c *configuration) SetupSmartContractAddresses(network string, smartContractAddresses *SmartContractAddresses) {
 	c.v.Set("networks."+network+".contractAddresses.identityFactory", smartContractAddresses.IdentityFactoryAddr)
 }
+
+func validateURL(u string) string {
+	parsedURL, err := url.Parse(u)
+	if err != nil {
+		log.Fatalf("error: %v", err)
```
Comment (vedhavyas) [inline config/configuration.go:699]: return an error and let the caller handle the error

File: config/configuration.go [hunk 3]
```diff
@@ -683,3 +692,20 @@ func CreateConfigFile(args map[string]interface{}) (*viper.Viper, error) {
...
+	}
+
+	if parsedURL.Scheme == "" {
+		parsedURL.Scheme = defaultURLScheme
+	}
+
+	if _, ok := allowedURLScheme[parsedURL.Scheme]; !ok {
+		log.Fatalf("error: url scheme %s is not allowed", parsedURL.Scheme)
```
Comment (vedhavyas) [inline config/configuration.go:707]: same as above :)

File: config/configuration.go [hunk 4]
```diff
@@ -30,6 +31,12 @@ import (
 
 var log = logging.Logger("config")
 
+var allowedURLScheme = map[string]struct{}{
+	"http":  {},
+	"https": {},
+	"ws":    {},
```
Comment (vedhavyas) [inline config/configuration.go:37]: we would need to include `wss` as well

**Comment by arberiii** [comment]
> Thank you for your contribution!!
> 
> The approach LGTM but we are missing this check while loading the config on every run. I think we could move this check into the config package and validate the checks there.

@vedhavyas Thank you for the review! I didn't understand where to move the validation as for now it is placed at go-centrifuge/config package

File: config/configuration.go [hunk 5]
```diff
@@ -683,3 +699,20 @@ func CreateConfigFile(args map[string]interface{}) (*viper.Viper, error) {
...
+	if err != nil {
+		return "", err
+	}
+
+	if parsedURL.Scheme == "" {
+		parsedURL.Scheme = defaultURLScheme
+	}
+
+	if _, ok := allowedURLScheme[parsedURL.Scheme]; !ok {
+		return "", errors.New("error: url scheme %s is not allowed", parsedURL.Scheme)
```
Comment (vedhavyas) [inline config/configuration.go:714]:
```suggestion
		return "", errors.New("url scheme %s is not allowed", parsedURL.Scheme)
```

**Comment by vedhavyas** [comment]
@arberiii I revert the previous validations as well since those validations happen when config file was created and `initializeViper` validations happens when created config is loaded. The reason we are doing this because once the config is created, I could also change the config file with `ftp://` instead. Let me know if this makes sense to you

**Comment by arberiii** [comment]
> @arberiii I would revert the previous validations as well since those validations happen when config file was created and `initializeViper` validations happens when created config is loaded. The reason we are doing this because once the config is created, I could also change the config file with `ftp://` instead. Let me know if this makes sense to you

Yes it makes sense. You actually explained well the first time, but my bad :). I revert it and fixed the assertion which was failing the tests

**Comment by arberiii** [comment]
I added 3 test cases with expected behaviors.

**Comment by vedhavyas** [comment]
@arberiii seems like `configuration_test.go` is not formatted properly
https://travis-ci.com/github/centrifuge/go-centrifuge/jobs/364988442#L718

**Comment by vedhavyas** [comment]
Closing this in favour of #1472

**Pull request closed without merging**